File: web/src/entities/lavaSea.ts
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { Resource } from '../resource.js';
import type { Animation } from '../media/animation.js';
import { ParticleSystem } from './particleSystem.js';

export class LavaSea extends Entity {
  private frame = 0;
  private safeMode = false;
  private safeLevel = 0;
  private readonly topAnim: Animation;
  private readonly fillAnim: Animation;

  constructor() {
    super();
    this.topAnim = Resource.getAnimation('data/images/lavatop.bmp', 2);
    this.fillAnim = Resource.getAnimation('data/images/lavafill.bmp', 2);
    this.setSize(new ImmutableVec2(10, 10));
  }

  getLayer(): number { return 4; }

  override getCollideTop(): number { return this.getCurrentY(); }
  override getCollideLeft(): number { return 0; }
  override getCollideRight(): number { return 100000; }
  override getCollideBottom(): number { return 100000; }

  private getCurrentY(): number {
    if (this.safeMode) return this.safeLevel;
    let sinwave = (Math.sin(this.frame / 100.0) + 1) / 2;
    sinwave = sinwave * sinwave * sinwave;
    return Math.round(this.getPosition().y - sinwave * 80 + 2);
  }

  override onLevelComplete(): void {
    this.safeLevel = this.getCurrentY();
    this.safeMode = true;
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const screenY = offsetY + this.getCurrentY();
    for (let x = -10 + offsetX % 10; x < 320; x += 10) {
      this.topAnim.drawFrame(ctx, Math.floor(this.frame / 30), x, screenY);
    }
    for (let x = -10 + offsetX % 10; x < 320; x += 10) {
      for (let y = screenY + 10; y < 240; y += 10) {
        this.fillAnim.drawFrame(ctx, Math.floor(this.frame / 30), x, y);
      }
    }
  }

  override update(): void {
    if (this.myRoom.getHero().Collides(this)) {
      this.myRoom.getHero().kill();
    }

    this.frame++;

    if (this.safeMode) {
      this.safeLevel++;
      if (this.safeLevel > this.getPosition().y + 10) {
        this.safeLevel = Math.round(this.getPosition().y + 10);
      }
    }

    if (Math.random() < 0.3) {
      const ps = new ParticleSystem(
        Resource.getAnimation('data/images/particles.bmp'),
        2, 30, 10, 1, 50, 10, new ImmutableVec2(0, -30), 5.0);
      const min = -this.myRoom.getCamera().getOffsetX();
      const max = 320 - this.myRoom.getCamera().getOffsetX();
      const rVal = Math.floor(Math.random() * (max - min)) + min;
      ps.setPositionWithSpread(new ImmutableVec2(rVal, this.getCurrentY()), 5, false);
      this.myRoom.addEntity(ps);
    }
  }
}

File: wii/src/entities/lava_sea.cpp
#include "entities/lava_sea.h"
#include "entities/particle_system.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "room.h"
#include "camera.h"
#include "constants.h"
#include <cmath>
#include <cstdlib>

LavaSea::LavaSea() {
    topAnim_ = Resource::getAnimation("data/images/lavatop.bmp", 2);
    fillAnim_ = Resource::getAnimation("data/images/lavafill.bmp", 2);
    setSize(Vec2(10, 10));
}

float LavaSea::getCurrentY() const {
    if (safeMode_) return safeLevel_;
    float sinwave = (std::sin(frame_ / 100.0f) + 1.0f) / 2.0f;
    sinwave = sinwave * sinwave * sinwave;
    return std::round(position_.y - sinwave * 80.0f + 2.0f);
}

void LavaSea::onLevelComplete() {
    safeLevel_ = getCurrentY();
    safeMode_ = true;
}

void LavaSea::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int screenY = offsetY + static_cast<int>(getCurrentY());
    int animFrame = (frame_ / 30);

    for (int x = -10 + offsetX % 10; x < SCREEN_WIDTH; x += 10) {
        topAnim_->drawFrame(renderer, animFrame, x, screenY);
    }
    for (int x = -10 + offsetX % 10; x < SCREEN_WIDTH; x += 10) {
        for (int y = screenY + 10; y < SCREEN_HEIGHT; y += 10) {
            fillAnim_->drawFrame(renderer, animFrame, x, y);
        }
    }
}

void LavaSea::update() {
    if (room_->getHero()->Collides(*this)) {
        room_->getHero()->kill();
    }

    frame_++;

    if (safeMode_) {
        safeLevel_ += 1.0f;
        if (safeLevel_ > position_.y + 10.0f) {
            safeLevel_ = std::round(position_.y + 10.0f);
        }
    }

    if (static_cast<float>(rand()) / RAND_MAX < 0.3f) {
        ParticleSystem* ps = new ParticleSystem(
            Resource::getAnimation("data/images/particles.bmp"),
            2, 30, 10, 1.0f, 50.0f, 10, Vec2(0, -30), 5.0f);
        float minX = static_cast<float>(-room_->getCamera()->getOffsetX());
        float maxX = static_cast<float>(SCREEN_WIDTH - room_->getCamera()->getOffsetX());
        float rVal = static_cast<float>(std::floor(
            static_cast<float>(rand()) / RAND_MAX * (maxX - minX) + minX));
        ps->setPositionWithSpread(Vec2(rVal, getCurrentY()), 5.0f, false);
        room_->addEntity(ps);
    }
}
